language/move-prover/spec-lang/src/ty.rs

```diff
@@ -233,7 +274,68 @@ impl Type {
             Vector(et) => et.is_incomplete(),
             Reference(_, bt) => bt.is_incomplete(),
             TypeDomain(bt) => bt.is_incomplete(),
-            Error | Primitive(..) | TypeParameter(..) | TypeLocal(..) => false,
+            Error | Primitive(..) | TypeLocal(..) | TypeParameter(_) => false,
+        }
+    }
+
+    pub fn is_open(&self) -> bool {
+        use Type::*;
+        match self {
+            Var(_) | TypeParameter(_) => true,
+            Tuple(ts) => ts.iter().any(|t| t.is_open()),
+            Fun(ts, r) => ts.iter().any(|t| t.is_open()) || r.is_open(),
+            Struct(_, _, ts) => ts.iter().any(|t| t.is_open()),
+            Vector(et) => et.is_open(),
+            Reference(_, bt) => bt.is_open(),
+            TypeDomain(bt) => bt.is_open(),
+            Error | Primitive(..) | TypeLocal(..) => true, // TODO: dunno about TypeLocal
```

Comment:
Did you meant to return `false` here? Currently this function will always return true. Also Primitive doesn't appear open.

TypeLocal should behave like TypeParameter. It's usage is as in `forall t: type: exists<R<t>>`.

You can treat Var and Error, but in fact after successful type check, those types must never appear. Perhaps just panic on them here and other functions which are not expected to be used during type checking.

Reply:
Subtly, the cases like `ts.iter().any(|t| t.is_open())` return `false` when `ts` is empty or consists of all closed types. But I was definitely wrong to return `true` for `Primitive`!

Will fix the TypeLocal, Var, and Err cases too--thanks!